src/main/java/mandel/citibike/map/MapController.java

```diff
@@ -0,0 +1,109 @@
+package mandel.citibike.map;
+
+import hu.akarnokd.rxjava3.swing.SwingSchedulers;
+import io.reactivex.rxjava3.disposables.Disposable;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import mandel.citibike.lambda.*;
+import org.jxmapviewer.JXMapViewer;
+import org.jxmapviewer.viewer.*;
+
+import java.awt.geom.Point2D;
+import java.util.*;
+import java.util.List;
+
+public class MapController {
+    private final MapFrame mapFrame;
+    private final JXMapViewer mapViewer;
+    private final RoutePainter routePainter;
+    private final WaypointPainter<Waypoint> waypointPainter;
+    private final MapComponent view;
+    private GeoPosition fromPosition = null;
+    private GeoPosition toPosition = null;
+    private GeoPosition start;
+    private GeoPosition end;
+
+    public MapController(MapFrame mapFrame, JXMapViewer mapViewer, RoutePainter routePainter,
```

Comment:
Instead of passing the frame, pass the text fields.

Right now the frame has the controller and the controller has the frame. That's circular referencing which is generally frowned upon.